pulser/simulation/simulation.py

```diff
@@ -442,23 +459,54 @@ def write_samples(
                                 )
                 self.samples["Local"][basis] = samples_dict
 
-    def _build_operator(
-        self, op_id: str, *qubit_ids: Union[str, int], global_op: bool = False
-    ) -> qutip.Qobj:
-        """Create qutip.Qobj with nontrivial action at *qubit_ids."""
-        if global_op:
-            return sum(
-                self._build_operator(op_id, q_id) for q_id in self._qdict
-            )
-        if len(set(qubit_ids)) < len(qubit_ids):
-            raise ValueError("Duplicate atom ids in argument list.")
-        # List of identity operators, except for op_id where requested:
-        op_list = [
-            self.op_matrix[op_id]
-            if j in map(self._qid_index.get, qubit_ids)
-            else self.op_matrix["I"]
-            for j in range(self._size)
-        ]
+    def build_operator(self, operations: Union[list, tuple]) -> qutip.Qobj:
+        """Creates an operator with non trivial actions on some qubits.
+
+        Takes as argument a list of tuples [(operator_1, qubits_1),
+        (operator_2, qubits_2)...]. Returns the operator given by the tensor
+        product of {operator_i applied on qubits_i} and Id on the rest.
+        (operator, 'global') returns the sum for all $j$ of operator
+        applied at qubit $j$ and identity elsewhere.
+
+        Example for 4 qubits: [(Z, [1, 2]), (Y, [3])] returns ZZYI
+        and [(X, 'global')] returns XIII + IXII + IIXI + IIIX
+
+        Args:
+            operations (list): List of tuples (operator, qubits)
+                operator can be a qutip.Quobj or a string key for
+                self.op_matrix qubits is the list on which operator
+                will be applied. The qubits can be passed as their
+                index or their label in the register.
+
+        Returns:
+            qutip.Qobj: the final operator.
+        """
+        op_list = [self.op_matrix["I"] for j in range(self._size)]
+
+        if not isinstance(operations, list):
+            operations = [operations]
+
+        for operator, qubits in operations:
+            if qubits == "global":
+                return sum(
+                    self.build_operator([(operator, [q_id])])
+                    for q_id in self._qdict
+                )
+            else:
+                if len(set(qubits)) < len(qubits):
+                    raise ValueError("Duplicate atom ids in argument list.")
+                if isinstance(operator, str):
+                    try:
+                        operator = self.op_matrix[operator]
+                    except KeyError:
+                        raise ValueError(f"{operator} is not a valid operator")
+                for qubit in qubits:
+                    try:
+                        k = self._qid_index[qubit]
+                    except KeyError:
+                        raise ValueError(f"{qubit} is not a valid qubit")
```

Comment:
Perhaps it's worth checking whether `set(qubits).issubset(set(self._qdict.keys()))` at the beginning, to avoid ever entering in the loop in the first place if there's a bad qubit

Reply:
I agree